File: data-structure-and-algorithm/java/peeking-iterator.java
import java.util.Iterator;

/**
 * Peeking Iterator
 * 
 * Given an Iterator class interface with methods: next() and hasNext(), design and implement a PeekingIterator that support the peek() operation --
 * it essentially peek() at the element that will be returned by the next call to next().
 * 
 * Example:
 * 
 * Assume that the iterator is initialized to the beginning of the list: [1,2,3].
 * 
 * Call next() gets you 1, the first element in the list.
 * Now you call peek() and it returns 2, the next element. Calling next() after that still return 2. 
 * You call next() the final time and it returns 3, the last element. 
 * Calling hasNext() after that should return false.
 * Hint:
 * 
 * Think of "looking ahead". You want to cache the next element.
 * Is one variable sufficient? Why or why not?
 * Test your design with call order of peek() before next() vs next() before peek().
 * For a clean implementation, check out Google's guava library source code.
 * Follow up: How would you extend your design to be generic and work with all types, not just integer?
 */

class PeekingIterator extends Iterator<Integer> {
    public PeekingIterator(Iterator<Integer> iterator) {
        this.iterator = new Iterator<Integer>(iterator);
        peekedValue = 0;
        hasPeeked = false;
    }

    public Integer peek() {
        if (!hasPeeked) {
            hasPeeked = true;
            peekedValue = super.next();
        }
        return peekedValue;
    }

    @Override
    public Integer next() {
        if (hasPeeked) {
            hasPeeked = false;
            return peekedValue;
        } 
        
        return iterator.next();
    }

    @Override
    public boolean hasNext() {
        return hasPeeked || iterator.hasNext();
    }

    private Iterator<Integer> iterator;
    private boolean hasPeeked ;
    private int peekedValue;
}

File: data-structure-and-algorithm/javascript/leetcode/peeking-iterator.js
/*
Design an iterator that supports the peek operation on a list in addition to the hasNext and the next operations.

Implement the PeekingIterator class:

PeekingIterator(int[] nums) Initializes the object with the given integer array nums.
int next() Returns the next element in the array and moves the pointer to the next element.
bool hasNext() Returns true if there are still elements in the array.
int peek() Returns the next element in the array without moving the pointer.
 

Example 1:

Input
["PeekingIterator", "next", "peek", "next", "next", "hasNext"]
[[[1, 2, 3]], [], [], [], [], []]
Output
[null, 1, 2, 2, 3, false]

Explanation
PeekingIterator peekingIterator = new PeekingIterator([1, 2, 3]); // [1,2,3]
peekingIterator.next();    // return 1, the pointer moves to the next element [1,2,3].
peekingIterator.peek();    // return 2, the pointer does not move [1,2,3].
peekingIterator.next();    // return 2, the pointer moves to the next element [1,2,3]
peekingIterator.next();    // return 3, the pointer moves to the next element [1,2,3]
peekingIterator.hasNext(); // return False
 

Constraints:

1 <= nums.length <= 1000
1 <= nums[i] <= 1000
All the calls to next and peek are valid.
At most 1000 calls will be made to next, hasNext, and peek.
 

Follow up: How would you extend your design to be generic and work with all types, not just integer?
*/

class PeekingIterator {
    constructor(iterator) {
        this.iterator = iterator
        this.value = 0
        this.hasPeeked = false
    }

    peek() {
        if (!this.hasPeeked) {
            this.value = this.iterator.next()
            this.hasPeeked = true
        }
        return this.value
    }

    hasNext() {
        return this.hasPeeked || this.iterator.hasNext()
    }

    next() {
        if (this.hasPeeked) {
            this.hasPeeked = false
            return this.value
        }

        return this.iterator.next()
    }
}